### PR TITLE
More Efficient Ordering of Shard Upload Execution

### PR DESCRIPTION
**Pull request by original-brownbear**
Inspired by https://github.com/elastic/elasticsearch/pull/39657 and resolving the issue raised in that PR:

The problem currently is that we are uploading all segments in a shard sequentially per shard and only parallelize snapshotting across shards but not within shards.
This PR adjusts the logic towards parallel segment uploads: 
* Instead of starting all shard uploads in parallel and without any order, run segment uploads shard by shard.
* Only acquire index commit when it's actually required (after determining the current state of a shard in the repository, not before that).
   * @DaveCTurner pointed out that going sequentially in determining the segments to upload shard by shard increases the timespan between index commits. This is true for the case pf having more snapshot threads than shards to be snapshotted. On the other hand if there's more shards to be snapshotted on a node than snapshot threads it's the other way around. Currently, in this situation once the snapshot thread pool is fully occupied a shard will have to wait for another shard to completely finish its snapshot before taking the index commit (which could be many minutes!) while now the time span between the index commits is likely never more than a few seconds apart (biggest difference between two index commit times is basically equal to the time it takes all the shard folder listing operations to finish) 
* Release index commit asap (with this change we generally way reduce the amount of time we hold on to an index commit in cases of uploading multiple shards with multiple segments)

### CODE REVIEW

**Comment by ywelsch** [comment]
@tlrx can you have a look here?

**Comment by tlrx** [comment]
@original-brownbear Do you think that this PR could be split into multiple smaller PRs? It is addressing several points (index commit acquiring/releasing, parallelization of uploads, factorization in a context class) and I must admit I have trouble to review this.

**Comment by original-brownbear** [comment]
@tlrx turns out this isn't so trivial to split up. Maybe to give you some background here take a look at https://github.com/elastic/elasticsearch/pull/42213 where we actively removed the index shard dependency from the repository.
Making a standalone change with the goal of 

>  I'd start with a simple change about where the indexcommit is acquired (after listing blobs) and released.

kind of tricky. So I'd have to bring the shard snapshot context abstraction back in isolation to achieve this I'm afraid.

The problem in general is, that the specific behavior change here (the shard upload ordering) is very small and merely a simple reordering of two for-loops being adjusted from parallel to non parallel and vice-versa. All the changes around it are merely there to enable doing this in a non-blocking manner (it would be a literal 10 line change if I could block inside the `snapshotShard` method).
The other thing is, moving the place where we acquire the shard index commit without the changes to parallelizing file uploads would actually make the problem addressed here worse I think. Now we'd wait even longer (because the whole blob listing IO is added to things ... even though it won't matter much practically) between acquiring two index commits and wouldn't get any faster release of those commits either.

Maybe it would be more helpful if I were to create a blocking version of just the shard upload reordering to make things clearer. We wouldn't be able to merge that as it would break the `SnapshotResiliencyTests` that rely on non of the snapshot methods blocking but I think that might make this easier to follow?
Another option would be to maybe just talk this through live next week? :)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

**Comment by original-brownbear** [comment]
Please hold off on reviewing this one, after the latest merge of `master` into this branch we seem to have picked up some conflict with SLM behaviour here and this change makes the following test run:

```
./gradlew :x-pack:plugin:ilm:qa:multi-node:integTestRunner -Dtests.seed=39DDC190C49A92E8 -Dtests.security.manager=true -Dtests.locale=kab-DZ -Dtests.timezone=Africa/Bamako -Dcompiler.java=12 -Druntime.java=11
```

flaky -> have to track this down first.

**Comment by original-brownbear** [comment]
Sort of false alarm here, the test failures are a result of https://github.com/elastic/elasticsearch/issues/21759 (the SLM test that failed uses a rate limit of 1 byte per second so if parallel writers hit that limit they can way overshoot that limit an make all the SNAPSHOT threads wait. I'll open a separate PR to improve the situation behind aborting snapshots, doing that here would way overcomplicate this PR.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java]: At first sight I liked this new class, but after iterating over this PR few times I like it a bit less. What I don't like here is that it adds another place where the `IndexShardSnapshotStatus` is modified (and kind of hide these modifications), and I also feel like it adds extra complexity.

Instead, I think we could wrap the IndexCommit into a AbstractRefCounted, increment it before triggering a file snapshot, decrement it once the snapshot of the file is done and let the IndexCommit be released itself once all file snapshots are done.

**Review by tlrx — COMMENTED** [review]
Sorry for the delay it took me to look at this. I left few comments I'd like to see addressed.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I figured this was a nicer way of handling the special case of `0` files than having a condition for the case of zero files and then using another level of indirection for `onResponse` in the grouped listener. Apparently not and conditional wins? :D

File: server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java]: > Instead, I think we could wrap the IndexCommit into a AbstractRefCounted, increment it before triggering a file snapshot, decrement it once the snapshot of the file is done and let the IndexCommit be released itself once all file snapshots are done.

That was my plan A at first. But then I started not liking that so much because:
* It makes the signature `org.elasticsearch.repositories.blobstore.BlobStoreRepository#snapshotShard` even more complex than it already is
* While it makes the state transitions of `IndexShardSnapshotStatus` a little more obvious now, it hides the connection between that ref counted instance, the index commit and the status doesn't it?

... Maybe it would be better to just flatten all this logic into `IndexShardSnapshotStatus` (or the logic of the shard status tracking into `ShardSnapshotContext`)? :)

The other thing about me adding this was that we will need to keep track of more state for each shard upload when fixing/optimizing the situation of unchanged shards in https://github.com/elastic/elasticsearch/issues/45736 ... my thinking was that if we change the signature like it was done here, we could do without yet another subsequent signature change across all repository implementations when dealing with that and simply add the necessary logic to convey an unchanged shard and its generation via `ShardSnapshotContext`?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: @tlrx yea that sounds like a reasonable approach for sure. The only issue with that, that I can see is this snippet in `snapshotFile`:

```
        } catch (Exception t) {
            failStoreIfCorrupted(store, t);
            snapshotStatus.addProcessedFile(0);
            throw t;
```

-> it is a little tricky to convert this into what you propose because we need the error but also the effect of a `0` return. Maybe we should make this nicer in a follow up? (also keeps the change complexity down a little to not change the abort mechanics here?)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java]: Done :)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java]: I just removed it, it doesn't apply in any way now.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: True, but maybe we should change it in another PR to not mix that change into this refactoring? (It was the same message before this change)

**Comment by original-brownbear** [comment]
@tlrx no worries with the delay, I know this is a nasty one -> thanks for taking a look! :) Tried addressing all the points you raised, let me know what you think.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: To be honest, I just find this confusing :(

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: The usage of snapshotStatus.addProcessedFile(0); should be the responsability of the caller, which should catch any exception.

+1 for a follow up

File: server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java]: It looks to me that we are trying to address different issues in this PR:
a) more efficient way to snapshot shard files
b) better visibility and better tracking of shard status and resources releases
c) prepare the work for #45736

I think this PR should be just about point a). I'm not saying that b) and c) are not valid and I'm pretty sure they deserve their own PRs :) but for now I don't think they should be an argument to introduce more complexity for what we want to solve here.

What we could do (sorry again if I wasn't clear before) is to acquire an index commit reference in  `SnapshotShardsService.snapshot()` like it was before, and ensure that the reference is released when the shard snapshot listener gets called. Then the listener is passed down to the `BlobStoreRepository#snapshotShard()`  which makes sure to call it through the group files listener. This way we don't touch how status is tracked and we don't introduce another complex object (which again, could be the subject of a dedicated PR). WDYT?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java]: @tlrx I think my main point above all the others was `c)` (sorry for not making that clear) and the fact that we would have to change the signature of `snapshotShard` twice (especially when those versions of the signature may be spread out across releases).

I think other than that, I just liked this approach because it kept the logic a little less spread out. But if the above point can't warm your heart to this change I can probably do what you're asking for pretty easily :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I will adjust this then once we settle on the other open point :)

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Kileld it -> now it's just a conditional here :)

File: server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java]: Done :) Now we're doing all the resource handling in the callback and are still creating the commit ref in the snapshot shards service like we did before.
Thanks for pushing on this ... turns out this massively simplified the change for the time being (200+ LoC changes saved :)) ... not sure why I did it so complicated the first time around.

**Comment by original-brownbear** [comment]
@tlrx I've now cleaned up all the things we talked about. This should be ready for another round of reviewing (of a much shorter diff :)).

File: server/src/main/java/org/elasticsearch/action/ActionListener.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/action/ActionListener.java]: typo: ia

File: server/src/main/java/org/elasticsearch/action/ActionListener.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/action/ActionListener.java]: Can you add a test for this?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java]: Maybe we could restore the comment about why we flush here?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I don't really like this `Runnable`, maybe we should use `StepListener`s instead?

I'm thinking of something like:
```java
final StepListener<Void> snapshotDone = new StepListener<>();
snapshotDone.whenComplete(listener::onResponse, onFailure); // snapshot is done, call parent listener

final StepListener<Void> snapshotFinalization = new StepListener<>();
snapshotFinalization.whenComplete(v -> {
    ... what is in the runnable...
    snapshotDone.onResponse(null);
}, snapshotDone::onFailre);

if (indexIncrementalFileCount == 0) {
    snapshotFinalization.onResponse(null);
} else {
    ... use the GroupedActionListener with a wrapped snapshotFinalization...
}
```

WDYT?

**Review by tlrx — COMMENTED** [review]
This looks great, thanks @original-brownbear! I prefer it this way.

I've left a suggestion about using StepListener, otherwise this is getting close.

File: server/src/main/java/org/elasticsearch/action/ActionListener.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/ActionListener.java]: Sure, I just added the same one we have for the after hook here as well :)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java]: Right, brought it back :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I like it :) Used a step listener the way you suggested now (with one naming change).

**Comment by original-brownbear** [comment]
Thanks @tlrx addressed all your points now I think :)

**Review by tlrx — APPROVED** [review]
LGTM

Thank you for considering my suggestions, I really like it now :) I think we should hold on the backport of this until 7.4 is cut.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: This is nice :+1:

**Comment by original-brownbear** [comment]
Thanks so much @tlrx! + def. this isn't `7.4` material :)

**Pull request merged**
